### ISSUE TITLE
schema.rb cannot be generated with existing migrations

### ISSUE THREAD

**Issue by TheNotary** [EFForg/action-center-platform#32]
Running `rake db:migrate` on a fresh database will not generate the same schema that is tracked into this git repo.  

# Arguments Towards Issue Resolution

As briefly discussed in person, my thoughts on this issue are that:

* It is unconventional to have a schema.rb that cannot be auto-generated via the rails migration feature
* It gives me a general feeling that things are buggy/ sloppy
* External (and possibly some internal) devs will have a slightly dis-similar database schema than our production server

I did a quick speed test and running `rake db:schema:load` takes 4 seconds, whereas `rake db:migrate` takes 5.  Given that this is a 1-time command in the development cycle, I perceive the performance difference to be negligible.  

# Precautions with Resolving this Issue

As you pointed out in person, I will make a point to:

* Not make changes to the production database
* Not hack that cool hidden rails table until I can find legitimate supporting documentation on doing so

# Difference between gh schema.rb and migration generated schema.rb

The difference between the gh schema.rb and the migration generated schema is very small.  Basically the gh version contains extra stuff that is not used or accessible to the rails app.  For the most part, these extra tables and fields exist because the migration files were created and run, and then deleted, but the schema.rb file was not reverted via `rake db:rollback`.  In a special case... the uuid thing...  we ran a migration to enable ossd on the database, but didn't backdate the timestamp on it, which prevented us from using the feature properly.  

What's different in the gh schema.rb:

* An unused table, 'action_page_images' exists, it's migration was eliminated because an alternate implementation was used instead
* An unused table, `locales` exists, I can't even find it's migration in the old git repo
*  The `action_pages` table contains an un-used `action_image` column with missing migration
*  The `action_pages` table contains an un-used `call_tool` boolean, with missing migration (we used call_campaign_id instead, btw)
*  The ID column for `ahoy_events` is defaulted to `uuid_generate_v4()` which we didn't set during the enable_ossd episode because we needed to backdate the enable_ossd migration to before the creation of the ahoy_events table (such that uuid_generate_v4 would work)
*  The ID column for `visits` is defaulted to `uuid_generate_v4`....

**Comment by TheNotary** [comment]
The resolution to these diffs involve:

* Backdating enable_ossd
* adding default uuid_generate_v4() lines to two migration files

The remaining dependencies can be handled by pulling in missing migrations and running the `db:migrate:down` command on your current dev setup (I feel we should also do this on the production database such that it's in-line with what is desirable --desirable meaning does not contain datatables and data fields that are unused by rails, and undocumented anywhere other than schema.rb but will respect your decision to not manipulate the production db unless necessary).  

The files needed for the transformation are here:  

The commands that conduct the database mutation are:

```
rake db:migrate:down VERSION=20140505141409  # removes table locales
rake db:migrate:down VERSION=20140504133807  # removes field action_page_image 
rake db:migrate:down VERSION=20140504122904  # removes table action_page_images
rake db:migrate:down VERSION=20140327210148  # field removes call_tool
```

After the commands are finished, schema.rb will be in a form that's more congruent with what the migrations are capable of generating (and your database will be corrected accordingly such that future migrations or `rake db:schema:dump` commands will return the same schema made possible by the migrations).  

After the magrate:down commands, the migrations should be deleted, they don't need to be tracked into version control since they are all pointless database changes that are undesirable.  `schema.rb` *should* be committed and pushed to upstream.  

In an accessory PR, I'll include the enable_ossd related migration patches which will complete the resolution of this issue.  

Thank you for your patients with me on this issue.  There will be vegan donuts on Monday from Pepples at sometime after 9:30 AM if everything goes according to my Monday morning plan :)

**Comment by TheNotary** [comment]
Here are the files that need to be dropped into `db/migrate/` and renamed from `.rb.txt` to `.rb` before those `db:migrate:down` commands will work.  

[20140327210148_add_call_tool_to_action_page.rb.txt](https://github.com/EFForg/action-center-platform/files/19896/20140327210148_add_call_tool_to_action_page.rb.txt)
[20140504122904_create_action_page_images.rb.txt](https://github.com/EFForg/action-center-platform/files/19893/20140504122904_create_action_page_images.rb.txt)
[20140504133807_add_action_page_image_to_action_page.rb.txt](https://github.com/EFForg/action-center-platform/files/19895/20140504133807_add_action_page_image_to_action_page.rb.txt)
[20140505141409_create_locales.rb.txt](https://github.com/EFForg/action-center-platform/files/19894/20140505141409_create_locales.rb.txt)

**Comment by TheNotary** [comment]
NOTE:  don't start running till Monday, I wanted to finish this Friday night but need to take a thought break.  


### Solution Overview
  * Clear out bunk tables and fields from migration (explained in this issue)
  * [easy-migration-fixes branch] Remove migrations AddGeoWorldMapTables and remove countries
  * [easy-migration-fixes branch] backdate enable_ossd and patch files to use 'uuid_generate_v4()'



# Procedure to Clear Out Bunk Tables and Fields from schema.rb

1)  git checkout -b fixing-migration-schema-discrepancy

2)  Copy/ paste sample migration files

3)  run migrate:down commands

```
rake db:migrate:down VERSION=20140505141409
rake db:migrate:down VERSION=20140504133807
rake db:migrate:down VERSION=20140504122904
rake db:migrate:down VERSION=20140327210148
```

4)  delete migration files

```
cd db/migrate
rm 20140327210148_add_call_tool_to_action_page.rb 20140504122904_create_action_page_images.rb 20140504133807_add_action_page_image_to_action_page.rb 20140505141409_create_locales.rb
```

5)  git add db/schema.rb; git commit -m "removed undesirable elements from schema.rb";

6)  push to github

7)  Never allow commiters to remove migrations from the repo again without first adjusting the db/schema.rb file (this took forever to sort out and we still have issues where production db will differ from some deploys)

**Comment by TheNotary** [comment]
Also wanted to leave this table so it's ready for Monday:

# Resolution Paths that Lead to schema.rb to be inline with migrations
(In order of least to most intrusive, I favor the 'intrusive' side because otherwise we're intentionally scripted pointless fields and tables in our schema.rb)

A) Peform easy-migration-fixes merge, and simply track in the erased migration scrips and leave them there

B) Pefrom Easy-migration-fixes, track in erased migrations, and write migrations to eliminate un-needed database elements

C)  Pefrom easy-migration-fixes, copy in erased migrations, perform `rake db:migrate:down` to fix schema.rb and your dev table so the un-needed tables don't show up in schema.rb next time you run `rake db:migrate`

I'm actually starting to favor B over C, since after you perform the migrations, we could theoretically delete the migrations that cancel each other at any point in the future.

**Comment by TheNotary** [comment]
this issue has been resolved with #33

**Comment by TheNotary** [comment]
This hasn't been pushed to production yet and needs to be tested on staging to ensure I don't bring act.eff.org offline ever again.
